Lib/test/test_buffer.py

```diff
@@ -4579,6 +4579,117 @@ def test_c_buffer(self):
+        self.assertEqual(rb_call_count, 1)
+
+    def test_inherit_but_return_something_else(self):
+        class A(bytearray):
+            def __buffer__(self, flags):
+                return memoryview(b"hello")
+
+        a = A(b"hello")
+        with memoryview(a) as mv:
+            self.assertEqual(mv.tobytes(), b"hello")
+
+        rb_call_count = 0
+        rb_raised = False
+        class B(bytearray):
+            def __buffer__(self, flags):
+                return memoryview(b"hello")
+            def __release_buffer__(self, view):
+                nonlocal rb_call_count
+                rb_call_count += 1
+                try:
+                    super().__release_buffer__(view)
+                except ValueError:
+                    nonlocal rb_raised
+                    rb_raised = True
+
+        b = B(b"hello")
+        with memoryview(b) as mv:
+            self.assertEqual(mv.tobytes(), b"hello")
+            self.assertEqual(rb_call_count, 0)
+        self.assertEqual(rb_call_count, 1)
+        self.assertIs(rb_raised, True)
+
+    def test_override_only_release(self):
+        class C(bytearray):
+            def __release_buffer__(self, buffer):
+                super().__release_buffer__(buffer)
+
+        c = C(b"hello")
+        with memoryview(c) as mv:
+            self.assertEqual(mv.tobytes(), b"hello")
+
+    def test_release_saves_reference(self):
+        smuggled_buffer = None
+
+        class C(bytearray):
+            def __release_buffer__(s, buffer: memoryview):
+                with self.assertRaises(ValueError):
+                    memoryview(buffer)
+                with self.assertRaises(ValueError):
+                    buffer.cast("b")
+                with self.assertRaises(ValueError):
+                    buffer.toreadonly()
+                with self.assertRaises(ValueError):
+                    buffer[:1]
+                with self.assertRaises(ValueError):
+                    buffer.__buffer__(0)
+                nonlocal smuggled_buffer
+                smuggled_buffer = buffer
+                self.assertEqual(buffer.tobytes(), b"hello")
+                super().__release_buffer__(buffer)
+
+        c = C(b"hello")
+        with memoryview(c) as mv:
+            self.assertEqual(mv.tobytes(), b"hello")
+        c.clear()
+        with self.assertRaises(ValueError):
+            smuggled_buffer.tobytes()
+
+    def test_release_saves_reference_no_subclassing(self):
+        ba = bytearray(b"hello")
+
+        class C:
+            def __buffer__(self, flags):
+                return memoryview(ba)
+
+            def __release_buffer__(self, buffer):
+                self.buffer = buffer
+
+        c = C()
+        with memoryview(c) as mv:
+            self.assertEqual(mv.tobytes(), b"hello")
+        self.assertEqual(c.buffer.tobytes(), b"hello")
+
+        with self.assertRaises(BufferError):
+            ba.clear()
+        c.buffer.release()
+        ba.clear()
+
```

Comment:
Can you add a test in which the class has multiple inheritance? Also tests for when there are two or classes in mro which  support buffer protocol?

Reply:
Adding some tests. It's hard to come up with a case where multiple bases support the C buffer protocol, because that will almost inevitably lead to:

```
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_buffer.py", line 4707, in test_two_buffer_bases
    class A(bytearray, bytes):
TypeError: multiple bases have instance lay-out conflict
```

Reply:
Hmm, maybe try one var length and one pure python type which implements buffer protocol.

Reply:
Yes, just pushed. Thanks for the review!

Reply:
>Adding some tests. It's hard to come up with a case where multiple bases support the C buffer protocol, because that will almost inevitably lead to:

I see, you would have to write a custom type in C whose layout doesn't conflicts for that. Probably an object which has just object header and supports buffer protocol.

Reply:
I am thinking of the case when in mro there are two classes which supports buffer protocol and you return the buffer created from calling (using super) the second one, in that case you need to call release function pointer of the second one not the first one, am I right?

Reply:
Thinking about it, that case probably would behave buggily; if we have another C class that implements `bf_getbuffer` but not `bf_releasebuffer` in the MRO before a class that implements both, we'd incorrectly call `bf_releasebuffer` on the second class. I'll write a test in `_testcapi`, verify, and fix. The fix should be to make `releasebuffer_maybe_call_super` do nothing if it encounters a class that has a C `bf_getbuffer` but no `bf_releasebuffer`.

Reply:
TBH, I agree this is an edge case but I try to be extra careful when touching `typeobject`, hope you don't mind the extra work.

Reply:
Better to catch this now then right before 3.12 final!

Reply:
Confirmed the bug I outlined above, I'll have to step out for a few hours and fix it after that.

Reply:
Thinking about it more, that crash isn't actually due to PEP 688: it reproduces without any Python `__buffer__` method. So I don't think there's any more interesting cases to cover for PEP 688.

Reply:
Filed #104297 for that case.